Allow arbitrary arguments to be passed to pip through pip.installed

Akm0d opened the pull request:
### What does this PR do?
Makes it possible to pass arbitrary arguments to pip from pip.installed

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/24751

### Previous Behavior
N/A

### New Behavior
 pip keyword and positional arguments not yet implemented in salt can be passed to pip through pip.installed

    .. code-block:: yaml

        pandas:
          pip.installed:
            - name: pandas
            - pip_future:
              - --latest-pip-kwarg:
                - param1
                - param2
              - --latest-pip-arg

    Will be translated into the following pip command:

    .. code-block:: bash

        pip install pandas --latest-pip-kwarg param1 --latest-pip-kwarg parm2 --latest-pip-arg

### Tests written?

Yes

### Commits signed with GPG?

Yes

waynew commented:
Uh... I think you might want to rebase your fixes :trollface: (or retarget your PR)

Akm0d committed:
Make it possible to pass arbitrary arguments to pip

twangboy commented:
@Akmod Looks like you have some lint

twangboy commented:
https://jenkinsci.saltstack.com/job/pr-lint/job/PR-52327/1/warnings52Result/new/

Akm0d committed:
Flattened nested logic

Akm0d committed:
Removed unittest call non-child class

Akm0d commented:
I don't understand why it's not passing the py3-centos-7 test.  I ran the tests manually using python3.6.1 on my own centos7 server and it all passed just fine

Akm0d commented:
Failing on Centos7 with python3 and windows, but not because of the commit.  Centos7 and Windows are having issues in Jenkins

Akm0d committed:
Merge branch 'develop' into fix_pip

waynew commented:
I'm not sure if the remaining problem is legit something with these tests - it is a problem in the pip state tests, so probably worth looking into 👍

Akm0d committed:
Merge branch 'develop' into fix_pip

Akm0d committed:
Merge branch 'develop' into fix_pip

waynew reviewed (approved):
👍

garethgreenaway committed:
Merge branch 'develop' into fix_pip

Akm0d committed:
Merge branch 'develop' into fix_pip

Akm0d committed:
Merge branch 'develop' into fix_pip

Akm0d committed:
Renamed pip_future and added documentation to pip_state
- Changed `pip_future` to `extra_args`
- Changed the execution module code block to a cli example
- Moved the state.sls example to pip_state.py

Akm0d committed:
Merge branch 'develop' of github.com:saltstack/salt into fix_pip

Akm0d committed:
Changed `pip_future` to `extra_args` in test cases
- solves https://github.com/saltstack/salt/issues/24751

twangboy commented on salt/modules/pip.py:
```diff
@@ -895,6 +915,24 @@ def install(pkgs=None,  # pylint: disable=R0912,R0913,R0914
 
+    if extra_args:
+        # These are arguments from the latest version of pip that
+        # have not yet been implemented in salt
```
`# These are any additional arguments to be passed as parameters to the pip command`

I don't think we need to specify that it's for arguments salt hasn't implemented yet

twangboy commented on salt/modules/pip.py:
```diff
@@ -612,6 +614,24 @@ def install(pkgs=None,  # pylint: disable=R0912,R0913,R0914
         Disable the cache.
 
+    extra_args
+        pip keyword and positional arguments not yet implemented in salt
```
`pip keyword and positional arguments to be passed as parameters to the pip command`

I don't think we need to denote `not yet implemented`

twangboy commented on salt/states/pip_state.py:
```diff
@@ -602,6 +605,23 @@ def installed(name,
                 - exists_action: i
 
+    extra_args
+        pip keyword and positional arguments not yet implemented in salt
```
same here

twangboy reviewed (changes requested):
A few minor things. Looks great.

Akm0d committed:
Merge branch 'develop' into fix_pip

Akm0d committed:
Merge branch 'develop' into fix_pip

Akm0d committed:
Merge branch 'develop' into fix_pip

max-arnold commented:
Can someone backport this to master branch?

Akm0d commented:
@max-arnold Done :+1:
